File: ClassWork/Trees/Heap/Heap.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<climits>
using namespace std;

class Heap
{
    public:
        vector<int> a;
        int sz=0;

    // ============================== Parent & Child Functions ==============================
        int parent(int x){
            return(x-1)/2;
        }

        int leftChild(int x){
            return 2*x+1;
        }
        int rightChild(int x){
            return 2*x+2;
        }
    // ============================== Heap Functions ==============================

        // ==================== PUSH ==================== it will automatically heapify. "heapify" means it will maintain the heap property
        void push(int val){
            int ind = sz;
            cout<<"Pushing at index : "<<ind<<endl;
            a.push_back(val);
            sz++;

            while(sz!=0){
                if(a[parent(ind)] < a[ind]){
                    swap(a[parent(ind)], a[ind]);
                    ind = parent(ind);
                }
                else{
                    break;
                }
            }           
        }

        // ==================== TOP ====================
        int top(){
            if(isEmpty()){
                return INT_MIN;
            }
            return a[0];
        }

        // ===================== POP ===================== only possible on the top
        void pop(){
            if(isEmpty()){
                cout<<"Empty"<<endl;
                return;
            }
            a[0] = a[sz-1];
            a.pop_back();
            sz--;
            int ind = 0;
            while(true){
                int curr = a[ind];
                int lc = leftChild(ind) < sz ? a[leftChild(ind)] : INT_MIN;
                int rc = rightChild(ind) < sz ? a[rightChild(ind)] : INT_MIN; 

                int maximum = max({curr, lc, rc});

                if(maximum == curr) break;
                if(maximum == lc){
                    cout<<"swapping it with left child = "<<lc<<endl;
                    swap(a[ind], a[leftChild(ind)]);
                    ind = leftChild(ind);
                }
                else{
                    cout<<"swapping it with right child = "<<rc<<endl;
                    swap(a[ind], a[rightChild(ind)]);
                    ind = rightChild(ind);
                }
            }
        }

        // ===================== EMPTY =====================
        bool isEmpty(){
            return size()==0;
        }

        int size(){
            return sz;
        }

        void display(){
            for(int x: a){
                cout<<x<<" -- ";
            }
            cout<<endl;
        }

} pq;

signed main(){
    cout<<"PUSH operation ================================================"<<endl;
    pq.push(90);
    pq.display();
    pq.push(50);
    pq.display();
    pq.push(70);
    pq.push(15);
    pq.push(35);
    pq.display();
    // cout<<"Till here no pop"<<endl;
    pq.push(75);
    pq.display();
    cout<<"POP operation ============================================="<<endl;
    pq.pop();
    pq.display();
    cout<<"Popping again"<<endl;
    pq.pop();
    pq.display();
}

File: ClassWork/Priority Queue & Heap/Heap/Heap.py
class Heap:
    def __init__(self):
        self.__l = []

    def __leftchild(self, p) -> int:
        return 2 * p + 1
    
    def __rightchild(self, p) -> int:
        return 2 * p + 2
    
    def __parent(self, p) -> int:
        return (p - 1) // 2

    def isEmpty(self) -> bool:
        return self.size() == 0
    
    def size(self) -> int:
        return len(self.__l)

    def push(self, value) -> int:  
        self.__l.append(value)
        index = len(self.__l) - 1

        while(index != 0 and self.__l[index] > self.__l[self.__parent(index)]):    # parent < child then swap 
            # pIndex = self.__parent(index)
            self.__l[index], self.__l[self.__parent(index)] = self.__l[self.__parent(index)], self.__l[index]

            index = self.__parent(index)

        return index

    def pop(self) -> int:
        if(self.isEmpty()):
            raise Exception("Heap is empty")
        
        self.__l[0], self.__l[-1] = self.__l[-1], self.__l[0]   # swap root and last

        value = self.__l.pop()   # pop last element
        index = 0

        while(self.__l[index] < self.__l[self.__leftchild(index)] or self.__l[index] < self.__l[self.__rightchild(index)]):
            curr, left, right = self.__l[index], -1, -1

            if(self.__l[self.__leftchild(index)] < len(self.__l)):   # if left child exists
                left = self.__l[self.__leftchild(index)]

            if(self.__l[self.__rightchild(index)] < len(self.__l)):  # if right child exists
                right = self.__l[self.__rightchild(index)]

            if(curr >= self.__l[self.__leftchild(index)] and curr >= self.__l[self.__rightchild(index)]):    # current is biggest
                break

            if(curr < left and left > right):  # left is bigger
                self.__l[index], self.__l[self.__leftchild(index)] = self.__l[self.__leftchild(index)], self.__l[index]
                index = self.__leftchild(index)  # go left
            else:    # right is bigger
                self.__l[index], self.__l[self.__rightchild(index)] = self.__l[self.__rightchild(index)], self.__l[index]
                index = self.__rightchild(index)  # go right

        return value


    def top(self) -> int:
        pass
    